designs/NuGet-PackageFileService.md

```diff
@@ -0,0 +1,112 @@
+
+# Spec Name
+
+* Status: In Review
+* Author(s): [Rob Relyea](https://github.com/rrelyea)
+* Issue: [10151](https://github.com/NuGet/Home/issues/10151) INuGetPackageFileService - Fetch Images and embedded licenses for codespaces and normal execution
+
+## Problem Background
+
+With Codespaces support added to NuGet client, our code needs to be able to handle all packages and communication with feeds happening on a server node. This is architected as async method calls to services that are remoted. In most cases, we also have chosen to have 1 code path, for standalone execution and Codespaces-connected execution. These services are still used in the standalone execution, however, the service runs in the same process, as opposed to being remoted across the network, as in the Codespaces-connected execution.
+
+Package Icons can either come from:
+- a package feeds search service
+- (old style) or a URI to an arbitrary place on the web (specified in the NuSpec via now deprecated [iconUrl](https://docs.microsoft.com/en-us/nuget/reference/nuspec#iconurl) element.)
+- embedded in a package (specified in the NuSpec via [icon](https://docs.microsoft.com/en-us/nuget/reference/nuspec#icon) element.)
+
+Licenses can also come from a few places:
+- a package feeds search service
+- a [licenseUrl](https://docs.microsoft.com/en-us/nuget/reference/nuspec#licenseurl) (deprecated)
+- a [license expression](https://docs.microsoft.com/en-us/nuget/reference/nuspec#license)
+- an [embedded license file](https://docs.microsoft.com/en-us/nuget/reference/nuspec#license) in the package
+
+## Who are the customers
+
+All PM UI customers.
+
+## Goals
+
+Support Codespaces-connected and standalone fetching of icons and licenses.
+
+## Non-Goals
+
+## Solution Overview
+
+### Fetching Files
+Defined a new interface for this service:
+
+```C#
+namespace NuGet.VisualStudio.Internal.Contracts
+{
+    public interface INuGetPackageFileService : IDisposable
+    {
+        ValueTask<Stream?> GetPackageIconAsync(PackageIdentity packageIdentity, CancellationToken cancellationToken);
+        ValueTask<Stream?> GetEmbeddedLicenseAsync(PackageIdentity packageIdentity, CancellationToken cancellationToken);

```

Comment:
Given that we expect licenses to be text, why not use a string return type.  If the license contents is very long, is there a perf benefit to using a stream?

Similarly, from an API design point of view, since we expect icons to be images, we could use that data type, but knowing that this is an RPC contract, I understand wanting to keep the byte[] -> image code out of the RPC converters.